File: backend/app/api/v1/analytics.py
from datetime import date
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import ValidationError
from sqlalchemy.orm import Session

from app.core.deps import TenantContext, require_permission
from app.db.session import get_db
from app.schemas.analytics import (
    AnalyticsActivitiesResponse,
    AnalyticsDealsResponse,
    AnalyticsForecastResponse,
    AnalyticsLeadsResponse,
    AnalyticsOverviewResponse,
    AnalyticsPipelineResponse,
    AnalyticsQueryParams,
    AnalyticsRange,
    AnalyticsRevenueResponse,
    AnalyticsTasksResponse,
    AnalyticsTeamResponse,
)
from app.schemas.dashboard import DashboardScope
from app.services.analytics_service import AnalyticsService

router = APIRouter(prefix="/tenants/{slug}/analytics", tags=["analytics"])


def parse_analytics_params(
    range: AnalyticsRange = Query(default=AnalyticsRange.last_30_days, alias="range"),
    start_date: date | None = Query(default=None),
    end_date: date | None = Query(default=None),
    scope: DashboardScope = Query(default=DashboardScope.my),
    timezone: str = Query(default="UTC", max_length=64),
    owner_id: UUID | None = Query(default=None),
    company_id: UUID | None = Query(default=None),
    stage: str | None = Query(default=None, max_length=32),
) -> AnalyticsQueryParams:
    try:
        return AnalyticsQueryParams(
            range=range,
            start_date=start_date,
            end_date=end_date,
            scope=scope,
            timezone=timezone,
            owner_id=owner_id,
            company_id=company_id,
            stage=stage,
        )
    except ValidationError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=exc.errors(),
        ) from exc


@router.get("/overview", response_model=AnalyticsOverviewResponse)
def get_analytics_overview(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("tenant:read")),
    db: Session = Depends(get_db),
) -> AnalyticsOverviewResponse:
    return AnalyticsService(db).get_overview(ctx, params)


@router.get("/revenue", response_model=AnalyticsRevenueResponse)
def get_analytics_revenue(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("deal:read")),
    db: Session = Depends(get_db),
) -> AnalyticsRevenueResponse:
    return AnalyticsService(db).get_revenue(ctx, params)


@router.get("/pipeline", response_model=AnalyticsPipelineResponse)
def get_analytics_pipeline(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("deal:read")),
    db: Session = Depends(get_db),
) -> AnalyticsPipelineResponse:
    return AnalyticsService(db).get_pipeline(ctx, params)


@router.get("/leads", response_model=AnalyticsLeadsResponse)
def get_analytics_leads(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("lead:read")),
    db: Session = Depends(get_db),
) -> AnalyticsLeadsResponse:
    return AnalyticsService(db).get_leads(ctx, params)


@router.get("/deals", response_model=AnalyticsDealsResponse)
def get_analytics_deals(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("deal:read")),
    db: Session = Depends(get_db),
) -> AnalyticsDealsResponse:
    return AnalyticsService(db).get_deals(ctx, params)


@router.get("/tasks", response_model=AnalyticsTasksResponse)
def get_analytics_tasks(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("task:read")),
    db: Session = Depends(get_db),
) -> AnalyticsTasksResponse:
    return AnalyticsService(db).get_tasks(ctx, params)


@router.get("/activities", response_model=AnalyticsActivitiesResponse)
def get_analytics_activities(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("activity:read")),
    db: Session = Depends(get_db),
) -> AnalyticsActivitiesResponse:
    return AnalyticsService(db).get_activities(ctx, params)


@router.get("/team", response_model=AnalyticsTeamResponse)
def get_analytics_team(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("user:read")),
    db: Session = Depends(get_db),
) -> AnalyticsTeamResponse:
    return AnalyticsService(db).get_team(ctx, params)


@router.get("/forecast", response_model=AnalyticsForecastResponse)
def get_analytics_forecast(
    params: AnalyticsQueryParams = Depends(parse_analytics_params),
    ctx: TenantContext = Depends(require_permission("deal:read")),
    db: Session = Depends(get_db),
) -> AnalyticsForecastResponse:
    return AnalyticsService(db).get_forecast(ctx, params)

File: frontend/lib/api/analytics.ts
import { apiFetch } from "@/lib/api/client";
import { getDefaultTimezone } from "@/lib/api/dashboard";
import type {
  AnalyticsActivitiesResponse,
  AnalyticsDealsResponse,
  AnalyticsFilters,
  AnalyticsForecastResponse,
  AnalyticsLeadsResponse,
  AnalyticsOverviewResponse,
  AnalyticsPipelineResponse,
  AnalyticsRevenueResponse,
  AnalyticsTasksResponse,
  AnalyticsTeamResponse,
} from "@/types/analytics";

function buildQuery(filters: AnalyticsFilters = {}): string {
  const params = new URLSearchParams();
  if (filters.range) params.set("range", filters.range);
  if (filters.scope) params.set("scope", filters.scope);
  if (filters.start_date) params.set("start_date", filters.start_date);
  if (filters.end_date) params.set("end_date", filters.end_date);
  params.set("timezone", filters.timezone ?? getDefaultTimezone());
  if (filters.owner_id) params.set("owner_id", filters.owner_id);
  if (filters.company_id) params.set("company_id", filters.company_id);
  if (filters.stage) params.set("stage", filters.stage);
  const qs = params.toString();
  return qs ? `?${qs}` : "";
}

export const ANALYTICS_RANGE_LABELS: Record<string, string> = {
  today: "Today",
  yesterday: "Yesterday",
  this_week: "This week",
  last_week: "Last week",
  last_7_days: "Last 7 days",
  last_30_days: "Last 30 days",
  this_month: "This month",
  last_month: "Last month",
  this_quarter: "This quarter",
  this_year: "This year",
  custom: "Custom range",
};

export async function getAnalyticsOverview(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsOverviewResponse> {
  return apiFetch<AnalyticsOverviewResponse>(`/tenants/${slug}/analytics/overview${buildQuery(filters)}`);
}

export async function getAnalyticsRevenue(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsRevenueResponse> {
  return apiFetch<AnalyticsRevenueResponse>(`/tenants/${slug}/analytics/revenue${buildQuery(filters)}`);
}

export async function getAnalyticsPipeline(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsPipelineResponse> {
  return apiFetch<AnalyticsPipelineResponse>(`/tenants/${slug}/analytics/pipeline${buildQuery(filters)}`);
}

export async function getAnalyticsLeads(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsLeadsResponse> {
  return apiFetch<AnalyticsLeadsResponse>(`/tenants/${slug}/analytics/leads${buildQuery(filters)}`);
}

export async function getAnalyticsDeals(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsDealsResponse> {
  return apiFetch<AnalyticsDealsResponse>(`/tenants/${slug}/analytics/deals${buildQuery(filters)}`);
}

export async function getAnalyticsTasks(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsTasksResponse> {
  return apiFetch<AnalyticsTasksResponse>(`/tenants/${slug}/analytics/tasks${buildQuery(filters)}`);
}

export async function getAnalyticsActivities(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsActivitiesResponse> {
  return apiFetch<AnalyticsActivitiesResponse>(`/tenants/${slug}/analytics/activities${buildQuery(filters)}`);
}

export async function getAnalyticsTeam(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsTeamResponse> {
  return apiFetch<AnalyticsTeamResponse>(`/tenants/${slug}/analytics/team${buildQuery(filters)}`);
}

export async function getAnalyticsForecast(
  slug: string,
  filters: AnalyticsFilters = {},
): Promise<AnalyticsForecastResponse> {
  return apiFetch<AnalyticsForecastResponse>(`/tenants/${slug}/analytics/forecast${buildQuery(filters)}`);
}
